### PR TITLE
Added C# Loop and Fibonacci benchmarks

### PR DESCRIPTION
**Pull request by Azrielx86**

### CODE REVIEW

**Comment by ShreyasJejurkar** [comment]
.NET 9 is generally available, so we should upgrade this to use that.

File: loops/csharp/csharp.csproj [hunk 1]
```diff
@@ -0,0 +1,10 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net8.0</TargetFramework>
```
Comment (amirvenus) [inline loops/csharp/csharp.csproj:5]: Please use .net 9

**Review by amirvenus — CHANGES_REQUESTED** [review]

File: loops/csharp/Program.cs
Comment (ebesirik) [inline loops/csharp/Program.cs:1]: I believe line 7 should be 100_000 instead of 10_000
(checked from Rust code)

File: loops/csharp/Program.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+﻿var u = int.Parse(args[0]);
+var r = Random.Shared.Next(0, 10_000);
+var a = new int[10_000];
+
+for (var i = 0; i < 10_000; i++)
+{
+    for (int j = 0; j < 10_000; j++)
```
Comment (ebesirik) [inline loops/csharp/Program.cs:7]: I believe line 7 should be 100_000 instead of 10_000
(checked from Rust code)

**Comment by ebesirik** [comment]
for the curious visitors like my self, these are my limited results with correct loop numbers;

C = 0.51s
C = 0.51s
C = 0.50s

Rust = 0.50s
Rust = 0.50s
Rust = 0.55s

Java = 0.62s
Java = 0.53s
Java = 0.54s

C# = 0.53s
C# = 0.52s
C# = 0.52s

**Comment by ebesirik** [comment]
> Anyone see slower time in net9? I tried in m3 macbook air
> 
> net9: C = 0.94s C = 0.93s C = 0.93s
> 
> Bun = 1.41s Bun = 1.40s Bun = 1.40s
> 
> C# = 1.73s C# = 1.78s C# = 1.72s
> 
> net8: C = 0.94s C = 0.93s C = 0.94s
> 
> Bun = 1.40s Bun = 1.40s Bun = 1.40s
> 
> C# = 0.97s C# = 0.97s C# = 0.97s

my tests were also ran on m3 air (.net8). Did you run it on low power mode @yunusefendi52 ?

**Comment by neon-sunset** [comment]
I can reproduce bad performance numbers on M1 Pro. Looking at C, C# and Go disasm - each arrives at its performance with different codegen.

C emits branchless loop body as expected, by far the best one:
```asm
LAB_100003ec8                                   XREF[1]:     100003f10(j)  
    mov        w11,#0x0
    mov        w10,#0x0
    ldr        w12,[x22, x8, LSL #0x2]=>local_9c88
LAB_100003ed4                                   XREF[1]:     100003ef8(j)  
    sdiv       w13,w11,w19
    msub       w13,w13,w19,w11
    add        w14,w11,#0x1
    sdiv       w15,w14,w19
    msub       w14,w15,w19,w14
    add        w12,w13,w12
    add        w10,w14,w10
    add        w11,w11,#0x2
    cmp        w11,w9
    b.ne       LAB_100003ed4
    add        w10,w10,w12
    add        w10,w10,w20
    str        w10,[x22, x8, LSL #0x2]=>local_9c88
    add        x8,x8,#0x1
    cmp        x8,x21
    b.ne       LAB_100003ec8
```

Meanwhile in C# the division is guarded with division by zero and overflow checks:
```asm
LAB_1000b3be8                                   XREF[1]:     1000b3c48(j)  
    mov        w3,wzr
    mov        w4,w1
LAB_1000b3bf0                                   XREF[1]:     1000b3c24(j)  
    ldr        w5,[x0, x4, LSL #0x2]
    cmp        w19,#0x0
    b.eq       LAB_1000b3c98
    cmn        w19,#0x1
    b.ne       LAB_1000b3c0c
    cmp        w3,#0x1
    b.vs       LAB_1000b3ca0
LAB_1000b3c0c                                   XREF[1]:     1000b3c00(j)  
    sdiv       w6,w3,w19
    msub       w6,w6,w19,w3
    add        w5,w5,w6
    str        w5,[x0, x4, LSL #0x2]
    add        w3,w3,#0x1
    cmp        w3,w2
    b.lt       LAB_1000b3bf0
    lsl        x3,x4,#0x2
    add        x3,x0,x3
    ldr        w4,[x3]
    add        w4,w4,w20
    str        w4,[x3]
    add        w1,w1,#0x1
    mov        w3,#0x2710
    cmp        w1,w3
    b.lt       LAB_1000b3be8
```
    
Go produces by far the worst codegen - even though the division itself has only single guard branch, the loop is "torn" and print + return block is placed in the middle of it. It takes too much space so I will omit it. .NET too sometimes exhibits such behavior under FullOpts (and, by extension, NativeAOT) compilation, but it doesn't happen often.

File: loops/csharp/csharp.csproj (hunk 1 above)
Comment (Azrielx86) [reply loops/csharp/csharp.csproj:5]: I didn't realize that .NET 9 is already in Gentoo, it is masked, but I'm installing it right now to update this :)

File: loops/csharp/Program.cs (hunk 2 above)
Comment (Azrielx86) [reply loops/csharp/Program.cs:7]: thanks! I didn't saw this, i'll fix it :)

**Comment by yunusefendi52** [comment]
@ebesirik Oh yap I just remember I am on low power mode on battery. But the result is still slower in net9 than in net8 or C

net9:
C = 0.70s
C = 0.50s
C = 0.51s

Bun = 1.00s
Bun = 0.80s
Bun = 0.81s

C# = 1.08s
C# = 0.99s
C# = 0.94s

net8:
C = 0.52s
C = 0.50s
C = 0.50s

Bun = 0.77s
Bun = 0.85s
Bun = 0.88s

C# = 0.55s
C# = 0.52s
C# = 0.54s

Which dotnet version should we use then?

**Comment by bddicken** [comment]
Thanks for submitting this, but we got c# in via a different PR so I am going to close this one out.

**Pull request closed without merging**